C# code generation breaks on resources with >2 levels of nested Resources

### Godot version

4.1.1

### System information

Godot v4.1.1.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 970 (NVIDIA; 31.0.15.3713) - Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz (8 Threads)

### Issue description

It appears nesting classes too deeply in a file breaks the code generator on C# projects (unsure if this is applicable to GDScript).

#### Problem Description

If you have a single file with deeply nested classes, the code generator can fail to build.

See reproduction steps for an example of what breaks it, as well as the log messages.

#### Expected Behavior

Ideally, you should be able to nest resources arbitrarily. Oftentimes, the reason a class is nested is because it will never be used elsewhere. The source generator should not care about the structure of my code, as long as it is valid code.

If this is something that should not be allowed, I would at the very least expect a build error message that conveys that. The errors does not make it clear what actionable steps can be taken to resolve the issue.

### Steps to reproduce

Attempt to build the following Resource:
```cs
using Godot;

public partial class Foo: Resource
{
  public partial class Bar: Resource
  {
    public partial class Baz: Resource
    {
      [Export]
      public string key;
      [Export]
      public bool value;
    }
    
    [Export]
    public Baz[] bazList;
  }
  
  [Export]
  public Bar[] barList;
}
```
It should fail with the following errors (`...` is meant to represent the project root dir):
```
D:\...\Godot.SourceGenerators\Godot.SourceGenerators.ScriptPropertiesGenerator\Foo.Bar.Baz_ScriptProperties.generated.cs(15,29): error CS0115: 'Bar.Foo.Baz.SetGodotClassPropertyValue(in godot_string_name, in godot_variant)': no suitable method found to override [D:\...\Example.csproj]
D:\...\Godot.SourceGenerators\Godot.SourceGenerators.ScriptPropertiesGenerator\Foo.Bar.Baz_ScriptProperties.generated.cs(27,29): error CS0115: 'Bar.Foo.Baz.GetGodotClassPropertyValue(in godot_string_name, out godot_variant)': no suitable method found to override [D:\...\Example.csproj]
D:\...\Godot.SourceGenerators\Godot.SourceGenerators.ScriptSerializationGenerator\Foo.Bar.Baz_ScriptSerialization.generated.cs(10,29): error CS0115: 'Bar.Foo.Baz.SaveGodotObjectData(GodotSerializationInfo)': no suitable method found to override [D:\...\Example.csproj]
D:\...\Godot.SourceGenerators\Godot.SourceGenerators.ScriptSerializationGenerator\Foo.Bar.Baz_ScriptSerialization.generated.cs(16,29): error CS0115: 'Bar.Foo.Baz.RestoreGodotObjectData(GodotSerializationInfo)': no suitable method found to override [D:\...\Example.csproj]
```
However, if you remove Foo, and simply nest once, this error does not occur despite the logic not changing.
```cs
using Godot;
public partial class Bar: Resource
{
  public partial class Baz: Resource
  {
    [Export]
    public string key;
    [Export]
    public bool value;
  }
  
  [Export]
  public Baz[] bazList;
}
```

### Minimal reproduction project

[Example.zip](https://github.com/godotengine/godot/files/12850933/Example.zip)

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Sample/NestedClass.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Sample/NestedClass.cs
@@ -1,0 +1,22 @@
+using System;
+
+namespace Godot.SourceGenerators.Sample;
+
+public partial class NestedClass : GodotObject
+{
+    public partial class NestedClass2 : GodotObject
+    {
+        public partial class NestedClass3 : GodotObject
+        {
+            [Signal]
+            public delegate void MySignalEventHandler(string str, int num);
+
+            [Export] private String field_String = "foo";
+            [Export] private String property_String { get; set; } = "foo";
+
+            private void Method()
+            {
+            }
+        }
+    }
+}

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs
@@ -105,16 +105,20 @@ namespace Godot.SourceGenerators
             if (isInnerClass)
             {
                 var containingType = symbol.ContainingType;
+                AppendPartialContainingTypeDeclarations(containingType);
 
-                while (containingType != null)
+                void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
                 {
+                    if (containingType == null)
+                        return;
+
+                    AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                     source.Append("partial ");
                     source.Append(containingType.GetDeclarationKeyword());
                     source.Append(" ");
                     source.Append(containingType.NameWithTypeParameters());
                     source.Append("\n{\n");
-
-                    containingType = containingType.ContainingType;
                 }
             }
 

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertiesGenerator.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertiesGenerator.cs
@@ -91,16 +91,20 @@ namespace Godot.SourceGenerators
             if (isInnerClass)
             {
                 var containingType = symbol.ContainingType;
+                AppendPartialContainingTypeDeclarations(containingType);
 
-                while (containingType != null)
+                void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
                 {
+                    if (containingType == null)
+                        return;
+
+                    AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                     source.Append("partial ");
                     source.Append(containingType.GetDeclarationKeyword());
                     source.Append(" ");
                     source.Append(containingType.NameWithTypeParameters());
                     source.Append("\n{\n");
-
-                    containingType = containingType.ContainingType;
                 }
             }
 

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptPropertyDefValGenerator.cs
@@ -88,16 +88,20 @@ namespace Godot.SourceGenerators
             if (isInnerClass)
             {
                 var containingType = symbol.ContainingType;
+                AppendPartialContainingTypeDeclarations(containingType);
 
-                while (containingType != null)
+                void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
                 {
+                    if (containingType == null)
+                        return;
+
+                    AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                     source.Append("partial ");
                     source.Append(containingType.GetDeclarationKeyword());
                     source.Append(" ");
                     source.Append(containingType.NameWithTypeParameters());
                     source.Append("\n{\n");
-
-                    containingType = containingType.ContainingType;
                 }
             }
 

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptSerializationGenerator.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptSerializationGenerator.cs
@@ -91,16 +91,20 @@ namespace Godot.SourceGenerators
             if (isInnerClass)
             {
                 var containingType = symbol.ContainingType;
+                AppendPartialContainingTypeDeclarations(containingType);
 
-                while (containingType != null)
+                void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
                 {
+                    if (containingType == null)
+                        return;
+
+                    AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                     source.Append("partial ");
                     source.Append(containingType.GetDeclarationKeyword());
                     source.Append(" ");
                     source.Append(containingType.NameWithTypeParameters());
                     source.Append("\n{\n");
-
-                    containingType = containingType.ContainingType;
                 }
             }
 

--- a/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptSignalsGenerator.cs
+++ b/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptSignalsGenerator.cs
@@ -100,16 +100,20 @@ namespace Godot.SourceGenerators
             if (isInnerClass)
             {
                 var containingType = symbol.ContainingType;
+                AppendPartialContainingTypeDeclarations(containingType);
 
-                while (containingType != null)
+                void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
                 {
+                    if (containingType == null)
+                        return;
+
+                    AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                     source.Append("partial ");
                     source.Append(containingType.GetDeclarationKeyword());
                     source.Append(" ");
                     source.Append(containingType.NameWithTypeParameters());
                     source.Append("\n{\n");
-
-                    containingType = containingType.ContainingType;
                 }
             }
 

--- a/modules/mono/glue/GodotSharp/Godot.SourceGenerators.Internal/UnmanagedCallbacksGenerator.cs
+++ b/modules/mono/glue/GodotSharp/Godot.SourceGenerators.Internal/UnmanagedCallbacksGenerator.cs
@@ -128,16 +128,20 @@ using Godot.NativeInterop;
         if (isInnerClass)
         {
             var containingType = symbol.ContainingType;
+            AppendPartialContainingTypeDeclarations(containingType);
 
-            while (containingType != null)
+            void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
             {
+                if (containingType == null)
+                    return;
+
+                AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                 source.Append("partial ");
                 source.Append(containingType.GetDeclarationKeyword());
                 source.Append(" ");
                 source.Append(containingType.NameWithTypeParameters());
                 source.Append("\n{\n");
-
-                containingType = containingType.ContainingType;
             }
         }
 
@@ -303,16 +307,20 @@ using Godot.NativeInterop;
         if (isInnerClass)
         {
             var containingType = symbol.ContainingType;
+            AppendPartialContainingTypeDeclarations(containingType);
 
-            while (containingType != null)
+            void AppendPartialContainingTypeDeclarations(INamedTypeSymbol? containingType)
             {
+                if (containingType == null)
+                    return;
+
+                AppendPartialContainingTypeDeclarations(containingType.ContainingType);
+
                 source.Append("partial ");
                 source.Append(containingType.GetDeclarationKeyword());
                 source.Append(" ");
                 source.Append(containingType.NameWithTypeParameters());
                 source.Append("\n{\n");
-
-                containingType = containingType.ContainingType;
             }
         }
 

